src/ring_theory/valuation/integers.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import ring_theory.valuation.basic
+
+/-!
+# Ring of integers under a given valuation
+
+The elements with valuation less than or equal to 1.
+
+TODO: Define characteristic predicate.
+-/
+
+universes u v
+
+namespace valuation
+
+section ring
+
+variables {R : Type u} {Γ₀ : Type v} [ring R] [linear_ordered_comm_group_with_zero Γ₀]
+variables (v : valuation R Γ₀)
+
+/-- The ring of integers under a given valuation is the subring of elements with valuation ≤ 1. -/
+def integer : subring R :=
+{ carrier := { x | v x ≤ 1 },
+  one_mem' := le_of_eq v.map_one,
+  mul_mem' := λ x y hx hy, trans_rel_right (≤) (v.map_mul x y) (mul_le_one' hx hy),
+  zero_mem' := trans_rel_right (≤) v.map_zero zero_le_one',
+  add_mem' := λ x y hx hy, le_trans (v.map_add x y)
+    (by { letI := classical.DLO Γ₀, exact max_le hx hy }),
+  neg_mem' := λ x hx, trans_rel_right (≤) (v.map_neg x) hx }
+
+namespace integer
+
+lemma one_of_is_unit (x : v.integer) (hx : is_unit x) : v x = 1 :=
```

Comment:
You aren't ever really treating it as a *sub*ring, right?
Would it make sense to set up a char.pred. immediately `valuation.is_ring_of_integers` and prove this for `integer`?
I guess it could assume `algebra ROI R` and demand that the algebra map is injective and that the range is exactly `v.integer`? A bit of a boring `char.pred`, but otherwise you'll end up sooner rather than later with a "ROI" that isn't `v.integer` and then you can't apply these lemmas.

Reply:
Why don't we merge this PR first and then set up the char pred later when we need it?

Reply:
Because you are already proving a bunch of stuff that could be done in terms of the char pred, right?